pypsa/optimization/constraints.py

```diff
@@ -415,40 +859,88 @@ def p_previous(idx):
+        )
+
+    # ext down
+    if not ramp_limit_down[ext_i].isnull().all().all():
+        p_nom = m[f"{c}-p_nom"].loc[ext_i]
+        limit_pu = DataArray(ramp_limit_down.reindex(active.index, columns=ext_i))
+        lhs = p_actual(ext_i) - p_previous(ext_i) + limit_pu * p_nom
+        rhs = rhs_start.reindex(columns=ext_i)
+        mask = active.reindex(columns=ext_i) & ~ramp_limit_down.isnull().reindex(
+            active.index, columns=ext_i
+        )
+        m.add_constraints(
+            lhs, ">=", rhs, name=f"{c}-ext-non-comm-{attr}-ramp_limit_down", mask=mask
         )
 
-    # ----------------------------- Extendable Generators ----------------------------- #
+    # ----------------------- Extendable, committable but non modular Generators ----------------------------- #
+
+    ext_i = n.get_extendable_i(c)
+    com_i = n.get_committable_i(c)
+    not_mod_i = n.df(c).query(f"({nominal_attrs[c]}_mod==0)").index
+    ext_i = ext_i.intersection(com_i).intersection(not_mod_i).rename(ext_i.name)
 
     assets = n.df(c).reindex(ext_i)
 
     # ext up
     if not ramp_limit_up[ext_i].isnull().all().all():
-        p_nom = m[f"{c}-p_nom"]
+        p_nom = m[f"{c}-p_nom"].loc[ext_i]
         limit_pu = DataArray(ramp_limit_up.reindex(active.index, columns=ext_i))
         lhs = p_actual(ext_i) - p_previous(ext_i) - limit_pu * p_nom
         rhs = rhs_start.reindex(columns=ext_i)
         mask = active.reindex(columns=ext_i) & ~ramp_limit_up.isnull().reindex(
             active.index, columns=ext_i
         )
         m.add_constraints(
-            lhs, "<=", rhs, name=f"{c}-ext-{attr}-ramp_limit_up", mask=mask
+            lhs, "<=", rhs, name=f"{c}-ext-comm-non-mod-{attr}-ramp_limit_up", mask=mask
         )
 
     # ext down
     if not ramp_limit_down[ext_i].isnull().all().all():
-        p_nom = m[f"{c}-p_nom"]
+        p_nom = m[f"{c}-p_nom"].loc[ext_i]
         limit_pu = DataArray(ramp_limit_down.reindex(active.index, columns=ext_i))
         lhs = p_actual(ext_i) - p_previous(ext_i) + limit_pu * p_nom
         rhs = rhs_start.reindex(columns=ext_i)
         mask = active.reindex(columns=ext_i) & ~ramp_limit_down.isnull().reindex(
             active.index, columns=ext_i
         )
         m.add_constraints(
-            lhs, ">=", rhs, name=f"{c}-ext-{attr}-ramp_limit_down", mask=mask
+            lhs,
+            ">=",
+            rhs,
+            name=f"{c}-ext-comm-non-mod-{attr}-ramp_limit_down",
+            mask=mask,
         )
 
-    # ----------------------------- Committable Generators ----------------------------- #
+    # ---------------- Committable but non extendable and non modular Generators ----------------------------- #
+
+    com_i = n.get_committable_i(c)
+    fix_i = n.get_non_extendable_i(c)
+    not_mod_i = n.df(c).query(f"({nominal_attrs[c]}_mod==0)").index
+    com_i = com_i.intersection(fix_i).intersection(not_mod_i).rename(com_i.name)
```

Comment:
No need to reeat this block so frequently; many lines are duplicated before; better to avoid duplication

Reply:
the same applies before and below